ocaml/xapi/xapi_vm_snapshot.ml

```diff
@@ -478,6 +479,36 @@ let revert_vm_fields ~__context ~snapshot ~vm =
     ~overrides ;
   TaskHelper.set_progress ~__context 0.1
 
+(* See CP-40528. A snapshot is a VM record that contains the record of
+   the VM at snapshot time in snapshot_metadata, including its VTPM.
+   Most data is restored from snapshot_metadata. At the time the
+   snapshot is taken, a copy of the VTPM is taken and this is part of
+   the VM record. We want to use a copy of this VTPM. We need to make a
+   copy because a snapshot can be restored multiple times and we have to
+   ensure the VTPM of the snapshot remains unaltered. *)
+
+let update_vtpm ~__context ~snapshot ~vm =
+  debug "%s for VM %s (1/2)" __FUNCTION__ (Ref.string_of vm) ;
+  let vtpms = Db.VM.get_VTPMs ~__context ~self:vm in
+  let uuids =
+    vtpms |> List.map (fun self -> Db.VTPM.get_uuid ~__context ~self)
+  in
+  (* destroy current VTPM *)
+  List.iter (fun vtpm -> Xapi_vtpm.destroy ~__context ~self:vtpm) vtpms ;
+  (* use copy of snapshot VTPM *)
+  let vtpms =
+    Db.VM.get_VTPMs ~__context ~self:snapshot
+    |> List.map (fun vtpm -> Xapi_vtpm.copy ~__context ~vM:vm vtpm)
+  in
+  (* reset UUIDs *)
+  List.combine vtpms uuids
+  |> List.iter (fun (vtpm, uuid) ->
```

Comment:
I think this makes two assumptions:
* That the current VM and its snapshot have the same number of VTPM objects.
* That the order of the VTPMs coming back from `Db.VM.get_VTPMs` is the same for the VM and for its snapshot.

I think that the first one in particular is not true in general, as one can easily add or remove a VTPM after creating a snapshot.

Reply:
Currently you can have at most one VTPM; but you are right that the current VM might have no VTPM - so this needs more careful coding; it might be necessary to remember the UUID of the VTPM that we snapshot as we otherwise might not be able to re-create it

Reply:
The code in `update_vifs_vbds_vgpus_and_vusbs` seems to make no attempt to preserve the UUID for the other devices upon reverting a VM.